File: 5_multiple_inheritance.py

# Multiple Inheritance 

class j:
    def displayj(self):
        print("i am j method")
    
class k:
    def displayk(self):
        print("i am k method")
class L(j , k):
    
    def __init__(self) -> None:
        super().__init__()
    
    def displayL(self):
        print("I am L method")
        
l = L()
l.displayj()
l.displayk()
l.displayL()































class person:
    def __init__(self,name,age) -> None:
        self.name = name
        self.age = age
    def personalInfo(self):
        print(self.name," - ",self.age)


class company:
    def __init__(self,cname,cloc) -> None:
        self.cloc = cloc
        self.cname = cname
    def companyInfo(self):
        print(self.cname," - ",self.cloc)


class employee(person,company):
    def __init__(self, name, age,cname,cloc) -> None:
        person.__init__(self,name,age)
        company.__init__(self,cname,cloc)

obj = employee("sayed",24,"Google","USA")
obj.companyInfo()
obj.personalInfo()

print(employee.mro()) #method resolution order

File: 5_multiple_inheritance.cpp
#include <iostream>
using namespace std;

class J {
public:
    void displayJ() {
        cout << "I am J method" << endl;
    }
};

class K {
public:
    void displayK() {
        cout << "I am K method" << endl;
    }
};

class L : public J, public K {  // Multiple inheritance: L inherits from both J and K
public:
    L() {
        // Constructor equivalent to Python's __init__
        // In C++, you don't need to explicitly call a constructor like 'super()'
    }

    void displayL() {
        cout << "I am L method" << endl;
    }
};

int main() {
    L l;  // Create an instance of class L
    l.displayJ();  // Call the method from class J
    l.displayK();  // Call the method from class K
    l.displayL();  // Call the method from class L

    return 0;
}
